File: games/gamesMicroservice.js
// gameMicroservice.js
const grpc = require('@grpc/grpc-js');
const protoLoader = require('@grpc/proto-loader');
// Charger le fichier game.proto
const gameProtoPath = 'games.proto';
const mongoose = require('mongoose');
const Games = require('./gamesModel');
const { Kafka } = require('kafkajs');

const gameProtoDefinition = protoLoader.loadSync(gameProtoPath, {
    keepCase: true,
    longs: String,
    enums: String,
    defaults: true,
    oneofs: true,
});

const kafka = new Kafka({
    clientId: 'my-app',
    brokers: ['localhost:9092']
});

//const producer = kafka.producer();

const gameProto = grpc.loadPackageDefinition(gameProtoDefinition).game;
const url = 'mongodb://0.0.0.0:27017/games';

mongoose.connect(url)
    .then(() => {
        console.log('connected to database!');
    }).catch((err) => {
        console.log(err);
    })

    const consumer = kafka.consumer({ groupId: 'game-group' });
    const runConsumer = async () => {
        try {
            await consumer.connect();
            console.log('Connected to Kafka');
            await consumer.subscribe({ topic: 'game_topic' });
            console.log('Subscribed to game_topic');
    
            await consumer.run({
                eachMessage: async ({ topic, partition, message }) => {
                    const { action, gameData } = JSON.parse(message.value.toString());
                    if (action === 'addGame') {
                        try {
                            const { title, description, type, prix } = gameData;
                            const newGame = new Games({ title, description, type, prix });
                            const savedGame = await newGame.save();
                            console.log('Added new game:', savedGame);
                        } catch (error) {
                            console.error('Error occurred while adding game:', error);
                        }
                    }
                    if (action === 'searchGames') {
                        try {
                            const games = await Games.find({}).exec();
                            console.log('Games : ', games);
                        } catch (error) {
                            console.error('Error occurred while fetching games:', error);
                        }
    
                    }
                    if (action === 'getGame') {
                        try {
                            const { gameId } = JSON.parse(message.value.toString());
                            const game = await Games.findOne({ _id: gameId }).exec();
                            console.log('Game : ', game);
                        } catch (error) {
                            console.error('Error occurred while fetching game:', error);
    
                        }
    
    
                    }
                },
            });
        } catch (error) {
            console.error('Error while setting up Kafka consumer:', error);
        }
    };
    
    runConsumer().catch(console.error);

const gameService = {
    getGame: async (call, callback) => {
        try {
            const gameId = call.request.game_id;
            //console.log(call.request);
            const game = await Games.findOne({ _id: gameId }).exec();

            if (!game) {
                callback({ code: grpc.status.NOT_FOUND, message: 'Game not found' });
                return;
            }
            callback(null, { game });

        } catch (error) {

            callback({ code: grpc.status.INTERNAL, message: 'Error occurred while fetching game' });
        }


    },
    searchGames: async (call, callback) => {
        try {
            const games = await Games.find({}).exec();


            callback(null, { games });
        } catch (error) {
 

            callback({ code: grpc.status.INTERNAL, message: 'Error occurred while fetching games' });
        }
    },
    
    addGame: async (call, callback) => {
        const { title, description,type,prix } = call.request;

        //console.log(call.request);
        const newGame = new Games({ title, description,type,prix });

        try {
         /*   await producer.connect();

            await producer.send({
                topic: 'tv-shows-topic',
                messages: [{ value: JSON.stringify(newGame) }],
            });

            await producer.disconnect();*/
          //  console.log(newGame);
            const savedGame = await newGame.save();

            callback(null, { game: savedGame });
        } catch (error) {
            callback({ code: grpc.status.INTERNAL, message: 'Error occurred while adding game' });
        }
    }


};



// Créer et démarrer le serveur gRPC
const server = new grpc.Server();
server.addService(gameProto.GameService.service, gameService);
const port = 50052;
server.bindAsync(`0.0.0.0:${port}`, grpc.ServerCredentials.createInsecure(),
    (err, port) => {
        if (err) {
            console.error('Échec de la liaison du serveur:', err);
            return;
        }
        console.log(`Le serveur s'exécute sur le port ${port}`);
        server.start();
    });
console.log(`Microservice de Jeux Vidéo en cours d'exécution sur le port
${port}`);

File: games/gamesMicroservice.py
import grpc
from concurrent import futures
import games_pb2
import games_pb2_grpc
from pymongo import MongoClient
from kafka import KafkaConsumer,KafkaProducer
import json
import logging
from bson import ObjectId
import threading

logging.basicConfig(level=logging.INFO)

class GameServiceServicer(games_pb2_grpc.GameServiceServicer):
    def __init__(self):
        self.client = MongoClient('mongodb://localhost:27017/')
        self.db = self.client['games']
        self.games_collection = self.db['games']
        logging.info('MongoDB connected')
        self.producer = KafkaProducer(bootstrap_servers='localhost:9092',
                                      value_serializer=lambda v: json.dumps(v).encode('utf-8'))
        logging.info('Kafka Producer init !')              

    def GetGame(self, request, context):
        game_id = request.game_id
        game = self.games_collection.find_one({'_id':ObjectId(game_id)})
        logging.info("Logging from gRPC!")
        if game:
            game_message = games_pb2.Game(id=str(game['_id']), title=game['title'], description=game['description'], type=game['type'], prix=game['prix'])
            self.send_message_to_kafka('getGame', {'gameId': game_id})
            return games_pb2.GetGameResponse(game=game_message)
        else:
            context.set_code(grpc.StatusCode.NOT_FOUND)
            self.send_message_to_kafka('getGame', {'gameId': 'null'})
            return games_pb2.GetGameResponse()

    def SearchGames(self, request, context):
        games = []
        logging.info("Logging from gRPC!")
        for game in self.games_collection.find():
            game_message = games_pb2.Game(id=str(game['_id']), title=game['title'], description=game['description'], type=game['type'], prix=game['prix'])
            games.append(game_message)
        return games_pb2.SearchGamesResponse(games=games)

    def AddGame(self, request, context):
        logging.info("Logging from gRPC!")
        new_game = {
            'title': request.title,
            'description': request.description,
            'type': request.type,
            'prix': request.prix
        }
        self.games_collection.insert_one(new_game)
        game_message = games_pb2.Game(id=str(new_game['_id']), title=new_game['title'], description=new_game['description'], type=new_game['type'], prix=new_game['prix'])
        return games_pb2.AddGameResponse(game=game_message)

    def send_message_to_kafka(self, action, data):
        message = {'action': action, **data}
        self.producer.send('games_orders_topic', value=message)
        logging.info(f"Message sent to Kafka: {message}")
    
    def start_kafka_consumer(self):
        consumer = KafkaConsumer(
            'game_topic',
            bootstrap_servers='localhost:9092',
            auto_offset_reset='earliest',
            group_id='game-group',
            client_id='projet',
            value_deserializer=lambda x: json.loads(x.decode('utf-8'))
        )

        for message in consumer:
            logging.info(message.value['action']) 
            logging.info('Logging from kafka!')
            if message.value['action'] == 'addGame':
                try:
                    game_data = message.value['gameData']
                    title = game_data.get('title')
                    description = game_data.get('description')
                    type_ = game_data.get('type')
                    prix = game_data.get('prix')
                    new_game = {
                        'title': title,
                        'description': description,
                        'type': type_,
                        'prix': prix
                    }            
                    saved_game = self.games_collection.insert_one(new_game)
                    logging.info('Added new game: %s', new_game)
                except Exception as e:
                    logging.error('Error occurred while adding game: %s', e)
            elif message.value['action'] == 'searchGames':
                try:
                    games = list(self.games_collection.find())
                    logging.info('Games: %s', games)
                except Exception as e:
                    logging.error('Error occurred while fetching games: %s', e)
            elif message.value['action'] == 'getGame':
                try:
                    game_id = message.value['gameId']
                    game = self.games_collection.find_one({'_id': ObjectId(game_id)})
                    self.send_message_to_kafka('getGame', {'gameId': game_id})
                    logging.info('Game: %s', game)
                except Exception as e:
                    logging.error('Error occurred while fetching game: %s', e)

def serve():
    logging.info('Starting gRPC server...')
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    games_pb2_grpc.add_GameServiceServicer_to_server(GameServiceServicer(), server)
    server.add_insecure_port('[::]:50052')
    server.start()
    logging.info('gRPC server started')
    server.wait_for_termination()

if __name__ == '__main__':
    game_service = GameServiceServicer()

    kafka_consumer_thread = threading.Thread(target=game_service.start_kafka_consumer)
    kafka_consumer_thread.start()

    serve()